### PR TITLE
GetMapLeaves in parallel

### LINKED ISSUE #1396: [Performance] Parallelize GetMapLeavesByRevision
`GetMapLeavesByRevision` has batch API semantics, but under-the-hood, the implementation is fetching each leaf sequentially.  For ~1000 leaves, this can take up to 10s. 

1. Explore fetching leaves in parallel. 
2. Explore caching common nodes between the leaves to avoid multiple db lookups of the same value.

### PR DESCRIPTION
**Pull request by gdbelvin**
Initial hacky attempt at #1396 

Use batch get methods for getting map leaves where we have them.
Where we don't have a batch method for fetching inclusion proofs, execute the fetches in parallel.

### CODE REVIEW

File: server/map_rpc_server.go [hunk 1]
```diff
@@ -128,45 +135,50 @@ func (t *TrillianMapServer) getLeavesByRevision(ctx context.Context, mapID int64
...
-			Inclusion: proof,
-		})
+	// Fetch inclusion proofs in parallel.
+	smtReader := merkle.NewSparseMerkleTreeReader(int64(mapRoot.Revision), hasher, tx)
+	inclusions := make([]*trillian.MapLeafInclusion, len(indices))
+	errs := make(chan error, len(indices))
+	var wg sync.WaitGroup
+	wg.Add(len(indices))
+	for i, index := range indices {
+		go func(i int, index []byte) {
```
Comment (pavelkalinnikov) [inline server/map_rpc_server.go]: I am concerned about 1 goroutine per index approach w.r.t. scale. Would introducing a batch request instead be more reliable?

File: server/map_rpc_server.go [hunk 2]
```diff
@@ -128,45 +135,50 @@ func (t *TrillianMapServer) getLeavesByRevision(ctx context.Context, mapID int64
...
+	}
 
-		inclusions = append(inclusions, &trillian.MapLeafInclusion{
-			Leaf:      leaf,
-			Inclusion: proof,
-		})
+	// Fetch inclusion proofs in parallel.
+	smtReader := merkle.NewSparseMerkleTreeReader(int64(mapRoot.Revision), hasher, tx)
+	inclusions := make([]*trillian.MapLeafInclusion, len(indices))
+	errs := make(chan error, len(indices))
```
[the remark this answers is not in the document]
Comment (gdbelvin) [reply server/map_rpc_server.go]: Do you think the extra code is worth the memory savings?

File: server/map_rpc_server.go (hunk 1 above)
Comment (gdbelvin) [reply server/map_rpc_server.go]: A batch index request would be infinitely better, but I'm working with what I've got.

Comment (pavelkalinnikov) [reply server/map_rpc_server.go]: Could you please leave a TODO to improve this?

File: server/map_rpc_server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply server/map_rpc_server.go]: Dunno, maybe not.

File: server/map_rpc_server.go (hunk 1 above)
Comment (pavelkalinnikov) [reply server/map_rpc_server.go]: Maybe even leave the issue open, to address it in a bit.

**Pull request merged**